File: app/integrations/bitrix24.py
from __future__ import annotations

import base64
from dataclasses import asdict
from datetime import datetime, timedelta
from typing import Any

import httpx

from app.config import Settings
from app.domain.models import BitrixSubmissionResult, LeadStatus, QualificationResult
from app.services.formatters import build_case_snapshot


class Bitrix24APIError(RuntimeError):
    pass


class Bitrix24Client:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings
        self._client = httpx.AsyncClient(timeout=20.0)

    async def close(self) -> None:
        await self._client.aclose()

    async def submit_intake(
        self,
        chat_id: int,
        answers: dict[str, str],
        qualification: QualificationResult,
    ) -> BitrixSubmissionResult:
        result = BitrixSubmissionResult()

        try:
            existing_lead_id = _int_or_none(answers.get("website_bitrix_lead_id"))
            if existing_lead_id:
                result.lead_id = existing_lead_id
                task_payload = self.build_task_payload(existing_lead_id, answers, qualification)
                task_response = await self.call("tasks.task.add", task_payload)
                result.task_id = self._extract_task_id(task_response)
                await self._try_start_task(result.task_id)

                timeline_payload = self.build_timeline_comment_payload(existing_lead_id, answers, qualification)
                timeline_response = await self.call("crm.timeline.comment.add", timeline_payload)
                result.timeline_comment_id = int(timeline_response)
                return result

            lead_payload = self.build_lead_payload(chat_id, answers, qualification)
            lead_response = await self.call("crm.item.add", lead_payload)
            result.lead_id = int(lead_response["item"]["id"])

            task_payload = self.build_task_payload(result.lead_id, answers, qualification)
            task_response = await self.call("tasks.task.add", task_payload)
            result.task_id = self._extract_task_id(task_response)
            await self._try_start_task(result.task_id)

            timeline_payload = self.build_timeline_comment_payload(result.lead_id, answers, qualification)
            timeline_response = await self.call("crm.timeline.comment.add", timeline_payload)
            result.timeline_comment_id = int(timeline_response)
        except Exception as exc:  # pragma: no cover - exercised through call-site fallbacks
            result.error_message = str(exc)

        return result

    async def call(self, method: str, payload: dict[str, Any]) -> Any:
        url = f"{self.settings.bitrix24_webhook_url.rstrip('/')}/{method}"
        response = await self._client.post(url, json=payload)
        response.raise_for_status()
        data = response.json()
        if "error" in data:
            raise Bitrix24APIError(f"{data['error']}: {data.get('error_description', '')}".strip())
        return data["result"]

    async def add_timeline_comment(
        self,
        lead_id: int,
        comment: str,
        files: list[tuple[str, bytes]] | None = None,
    ) -> int | None:
        fields: dict[str, Any] = {
            "ENTITY_ID": lead_id,
            "ENTITY_TYPE": "lead",
            "COMMENT": comment,
        }
        if files:
            fields["FILES"] = [
                [file_name, base64.b64encode(content).decode("ascii")]
                for file_name, content in files
            ]

        response = await self.call(
            "crm.timeline.comment.add",
            {"fields": fields},
        )
        return int(response) if response else None

    async def _try_start_task(self, task_id: int | None) -> None:
        if not task_id:
            return
        try:
            await self.call("tasks.task.start", {"taskId": task_id})
        except Exception:
            return

    def build_lead_payload(
        self,
        chat_id: int,
        answers: dict[str, str],
        qualification: QualificationResult,
    ) -> dict[str, Any]:
        first_name, last_name = _split_name(answers.get("full_name", ""))
        fields: dict[str, Any] = {
            "title": f"{answers.get('practice_area', 'Юр. запрос')} | {answers.get('full_name', 'Без имени')}",
            "name": first_name,
            "lastName": last_name,
            "sourceId": self.settings.bitrix24_source_id,
            "sourceDescription": "Telegram legal intake bot",
            "assignedById": qualification.responsible_id or self.settings.bitrix24_responsible_id,
            "comments": build_case_snapshot(answers, qualification),
            "statusDescription": self._build_status_description(answers, qualification),
            "statusId": qualification.bitrix_status_id or self.settings.bitrix24_target_status_id,
            "originatorId": "telegram-bot",
            "originId": str(chat_id),
            "utmSource": "telegram",
            "utmMedium": "bot",
            "companyTitle": answers.get("client_type") if answers.get("client_type") == "Компания" else None,
            "fm": [
                {
                    "typeId": "PHONE",
                    "valueType": "WORK",
                    "value": answers.get("phone", ""),
                }
            ],
        }
        email = answers.get("email", "")
        if email and email.lower() != "нет":
            fields["fm"].append(
                {
                    "typeId": "EMAIL",
                    "valueType": "WORK",
                    "value": email,
                }
            )
        return {
            "entityTypeId": self.settings.bitrix24_entity_type_id,
            "fields": fields,
        }

    def build_task_payload(
        self,
        lead_id: int,
        answers: dict[str, str],
        qualification: QualificationResult,
    ) -> dict[str, Any]:
        deadline = datetime.now().astimezone() + timedelta(hours=qualification.recommended_response_hours)
        priority = 2 if qualification.status is LeadStatus.HOT else 1
        return {
            "fields": {
                "TITLE": (
                    f"[{answers.get('preferred_time', 'слот не выбран')}] "
                    f"[{qualification.manager_priority}] "
                    f"{answers.get('practice_area', 'Юр. лид')} | {answers.get('full_name', 'Без имени')}"
                ),
                "DESCRIPTION": build_case_snapshot(answers, qualification),
                "CREATED_BY": self.settings.creator_id,
                "RESPONSIBLE_ID": qualification.responsible_id or self.settings.bitrix24_responsible_id,
                "PRIORITY": priority,
                "DEADLINE": deadline.replace(microsecond=0).isoformat(),
                "UF_CRM_TASK": [f"L_{lead_id}"],
            }
        }

    def build_timeline_comment_payload(
        self,
        lead_id: int,
        answers: dict[str, str],
        qualification: QualificationResult,
    ) -> dict[str, Any]:
        return {
            "fields": {
                "ENTITY_ID": lead_id,
                "ENTITY_TYPE": "lead",
                "COMMENT": build_case_snapshot(answers, qualification),
            }
        }

    @staticmethod
    def _extract_task_id(task_response: dict[str, Any]) -> int | None:
        if "task" in task_response and isinstance(task_response["task"], dict):
            task_id = task_response["task"].get("id")
            return int(task_id) if task_id else None
        if "item" in task_response and isinstance(task_response["item"], dict):
            task_id = task_response["item"].get("id")
            return int(task_id) if task_id else None
        return None

    @staticmethod
    def _build_status_description(
        answers: dict[str, str],
        qualification: QualificationResult,
    ) -> str:
        fit = {
            "target": "Целевой лид",
            "adjacent": "Смежное обращение",
            "non_target": "Нецелевое обращение",
        }.get(qualification.fit.value, "Целевой лид")
        route = {
            "bankruptcy": "банкротство",
            "family": "семейное право",
            "manual_review": "ручная проверка",
        }.get(qualification.route, "общий маршрут")
        return f"{fit}. Маршрут: {route}. Контакт: {answers.get('preferred_time', '-')}."


def _split_name(full_name: str) -> tuple[str, str]:
    parts = [part for part in full_name.split() if part]
    if not parts:
        return "", ""
    if len(parts) == 1:
        return parts[0], ""
    return parts[0], " ".join(parts[1:])


def _int_or_none(value: object) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None

File: site/lib/bitrix24.js
import { services } from "./site-data.js";

const DEFAULT_ENTITY_TYPE_ID = 1;

export function isBitrixConfigured() {
  return Boolean(process.env.BITRIX24_WEBHOOK_URL);
}

export async function submitLeadToBitrix(lead) {
  if (!isBitrixConfigured()) {
    return {
      configured: false,
      leadId: null,
      taskId: null,
      timelineCommentId: null,
      error: null
    };
  }

  const service = services.find((item) => item.slug === lead.service);
  const leadResponse = await callBitrix("crm.item.add", buildLeadPayload(lead, service));
  const leadId = Number(leadResponse?.item?.id);

  if (!leadId) {
    throw new Error("Bitrix24 did not return created lead id.");
  }

  const taskResponse = await callBitrix("tasks.task.add", buildTaskPayload(lead, service, leadId));
  const taskId = Number(taskResponse?.task?.id || taskResponse?.item?.id || 0) || null;
  const taskStarted = taskId ? await startTask(taskId) : false;

  const timelineResponse = await callBitrix(
    "crm.timeline.comment.add",
    buildTimelinePayload(lead, service, leadId)
  );
  const timelineCommentId = Number(timelineResponse || 0) || null;

  return {
    configured: true,
    leadId,
    taskId,
    taskStarted,
    timelineCommentId,
    error: null
  };
}

async function callBitrix(method, payload) {
  const baseUrl = process.env.BITRIX24_WEBHOOK_URL.replace(/\/+$/, "");
  let response;

  try {
    response = await fetch(`${baseUrl}/${method}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(payload),
      cache: "no-store"
    });
  } catch (error) {
    const cause = error?.cause?.message ? ` (${error.cause.message})` : "";
    throw new Error(`Bitrix24 network error: ${error.message}${cause}`);
  }

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(`Bitrix24 HTTP ${response.status}: ${JSON.stringify(data)}`);
  }

  if (data?.error) {
    throw new Error(`${data.error}: ${data.error_description || "Bitrix24 API error"}`);
  }

  return data?.result;
}

async function startTask(taskId) {
  try {
    await callBitrix("tasks.task.start", { taskId });
    return true;
  } catch {
    return false;
  }
}

function buildLeadPayload(lead, service) {
  const [firstName, ...restName] = String(lead.name || "").trim().split(/\s+/);
  const fields = {
    title: `${service?.shortTitle || "Юридическая заявка"} | ${lead.name}`,
    name: firstName || lead.name,
    lastName: restName.join(" "),
    sourceId: process.env.BITRIX24_SOURCE_ID || "WEB",
    sourceDescription: "Website diagnostic form",
    assignedById: Number(lead.routing?.responsibleId || process.env.BITRIX24_RESPONSIBLE_ID || 1),
    comments: buildCrmComment(lead, service),
    statusDescription: buildStatusDescription(lead),
    statusId: lead.routing?.statusId || process.env.BITRIX24_TARGET_STATUS_ID || "IN_PROCESS",
    originatorId: "legal-site",
    originId: lead.requestId,
    utmSource: lead.utm?.utm_source || "website",
    utmMedium: lead.utm?.utm_medium || "form",
    utmCampaign: lead.utm?.utm_campaign || "",
    fm: [
      {
        typeId: "PHONE",
        valueType: "WORK",
        value: lead.phone
      }
    ]
  };

  return {
    entityTypeId: Number(process.env.BITRIX24_ENTITY_TYPE_ID || DEFAULT_ENTITY_TYPE_ID),
    fields
  };
}

function buildTaskPayload(lead, service, leadId) {
  const deadline = new Date(Date.now() + getResponseWindowHours(lead.urgency) * 60 * 60 * 1000);
  return {
    fields: {
      TITLE: `[${lead.consultationSlot || "слот не выбран"}] [website] ${service?.shortTitle || "Юридическая заявка"} | ${lead.name}`,
      DESCRIPTION: buildTaskDescription(lead, service),
      CREATED_BY: Number(
        process.env.BITRIX24_CREATOR_ID || process.env.BITRIX24_RESPONSIBLE_ID || 1
      ),
      RESPONSIBLE_ID: Number(lead.routing?.responsibleId || process.env.BITRIX24_RESPONSIBLE_ID || 1),
      PRIORITY: getTaskPriority(lead),
      DEADLINE: deadline.toISOString(),
      UF_CRM_TASK: [`L_${leadId}`]
    }
  };
}

function buildTimelinePayload(lead, service, leadId) {
  return {
    fields: {
      ENTITY_ID: leadId,
      ENTITY_TYPE: "lead",
      COMMENT: buildCrmComment(lead, service)
    }
  };
}

export function buildCrmComment(lead, service) {
  return [
    `Направление: ${service?.shortTitle || lead.service}`,
    `Клиент: ${lead.name}`,
    `Телефон: ${lead.phone}`,
    `Предпочтительное время: ${lead.consultationSlot || "-"}`,
    `Канал связи: ${lead.contact}`,
    `Срочность: ${lead.urgency}`,
    `Регион: ${lead.region || "-"}`,
    `Документы: ${lead.documents || "-"}`,
    `Ключевые факты: ${lead.caseDetails || "-"}`,
    "",
    "Ситуация:",
    cleanSituation(lead.situation)
  ].join("\n");
}

function buildTaskDescription(lead, service) {
  return [
    buildCrmComment(lead, service),
    "",
    `Маршрут: ${humanFit(lead.routing?.fit)}. ${humanRoute(lead.routing?.route)}.`
  ].join("\n");
}

function cleanSituation(situation = "") {
  return String(situation)
    .split(/\r?\n/)
    .filter((line) => !line.startsWith("Направление:"))
    .filter((line) => !line.startsWith("Ключевые факты:"))
    .filter((line) => !line.startsWith("Документы:"))
    .filter((line) => !line.startsWith("Регион:"))
    .join("\n")
    .trim();
}

function getResponseWindowHours(urgency) {
  if (String(urgency || "").includes("Сегодня")) return 1;
  if (String(urgency || "").includes("3")) return 4;
  return 24;
}

function getTaskPriority(lead) {
  if (lead.routing?.fit === "non_target") return 1;
  if (String(lead.urgency || "").includes("Сегодня")) return 2;
  if (String(lead.urgency || "").includes("3")) return 2;
  return 1;
}

function buildStatusDescription(lead) {
  return `${humanFit(lead.routing?.fit)}. ${humanRoute(lead.routing?.route)}. Контакт: ${
    lead.consultationSlot || "-"
  }.`;
}

function humanFit(fit) {
  if (fit === "non_target") return "Нецелевое обращение";
  if (fit === "adjacent") return "Смежное обращение";
  return "Целевой лид";
}

function humanRoute(route) {
  if (route === "bankruptcy") return "Маршрут: банкротство";
  if (route === "family") return "Маршрут: семейное право";
  if (route === "manual_review") return "Маршрут: ручная проверка";
  return "Маршрут: общий";
}
